quickwit/quickwit-search/src/leaf.rs

```diff
@@ -829,6 +829,104 @@ impl CanSplitDoBetter {
     }
 }
 
+/// `multi_leaf_search` searches multiple indices and multiple splits.
+#[instrument(skip_all, fields(index = ?leaf_search_request.search_request.as_ref().unwrap().index_id_patterns))]
+pub async fn multi_leaf_search(
+    searcher_context: Arc<SearcherContext>,
+    leaf_search_request: LeafSearchRequest,
+    storage_resolver: &StorageResolver,
+) -> Result<LeafSearchResponse, SearchError> {
+    let search_request: Arc<SearchRequest> = leaf_search_request
+        .search_request
+        .ok_or_else(|| SearchError::Internal("no search request".to_string()))?
+        .into();
+
+    let doc_mappers: Vec<Arc<dyn DocMapper>> = leaf_search_request
+        .doc_mappers
+        .iter()
+        .map(|doc_mapper| deserialize_doc_mapper(doc_mapper))
+        .collect::<crate::Result<_>>()?;
+    // Creates a collector which merges responses into one
+    let aggregation_limits = searcher_context.create_new_aggregation_limits();
+    // TODO: to avoid lockstep, we should pull up the future creation over the list of split ids
+    // and have the semaphore on this level.
+    // This will lower resource consumption due to less in-flight futures and avoid contention.
+    // It also allows passing early exit conditions between indices.
+    //
+    // It is a little bit tricky how to handle which is now the incremental_merge_collector, one
+    // per index, e.g. when to merge results and how to avoid lock contention.
+    let mut leaf_request_tasks = Vec::new();
+
+    for leaf_search_request_ref in leaf_search_request.leaf_requests.into_iter() {
+        let index_uri = quickwit_common::uri::Uri::from_str(
+            &leaf_search_request.index_uris[leaf_search_request_ref.index_uri_ord as usize],
+        )?;
+        let doc_mapper = doc_mappers[leaf_search_request_ref.doc_mapper_ord as usize].clone();
+
+        let leaf_request_future = tokio::spawn(
+            resolve_storage_and_leaf_search(
+                searcher_context.clone(),
+                search_request.clone(),
+                index_uri,
+                storage_resolver.clone(),
+                leaf_search_request_ref.split_offsets,
+                doc_mapper,
+                aggregation_limits.clone(),
+            )
+            .in_current_span(),
+        );
+        leaf_request_tasks.push(leaf_request_future);
+    }
+
+    let leaf_responses = try_join_all(leaf_request_tasks).await?;
+    let merge_collector = make_merge_collector(&search_request, &aggregation_limits)?;
+    let mut incremental_merge_collector = IncrementalCollector::new(merge_collector);
+    for result in leaf_responses {
+        match result {
+            Ok(result) => {
+                incremental_merge_collector.add_result(result)?;
```

Comment:
ah add_result is light anyway... maybe it does not matter that much as long as we don't have any lazy optimization.